django/core/files/storage/memory.py

```diff
@@ -0,0 +1,243 @@
+            )
+            if current_node is None:
+                break
+
+        if current_node is None and check_exists:
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), path)
+
+        # If a leaf_cls is not None, check if leaf node is of right type
+        if leaf_cls and not isinstance(current_node, leaf_cls):
+            error_cls, error_code = (
+                (NotADirectoryError, errno.ENOTDIR)
+                if leaf_cls is InMemoryDirNode
+                else (IsADirectoryError, errno.EISDIR)
+            )
+            raise error_cls(error_code, os.strerror(error_code), path)
+
+        return current_node
+
+    def _resolve_child(self, path_segment, create_if_missing, child_cls):
+        if path_segment in ("", "."):
+            return self
+        if create_if_missing:
+            self._update_accessed_time()
+            self._update_modified_time()
+            return self._children.setdefault(path_segment, child_cls())
+        return self._children.get(path_segment)
+
+    def list_dir_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryDirNode)]
+
+    def list_file_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryFileNode)]
+
+    def remove_child(self, name):
+        if name in self._children:
+            self._update_accessed_time()
+            self._update_modified_time()
+            del self._children[name]
+
+
+@deconstructible(path="django.core.files.storage.InMemoryStorage")
+class InMemoryStorage(Storage, StorageSettingsMixin):
+    """
+    A storage saving files in memory
+    """
+
+    def __init__(self, base_url=None):
+        self._root = InMemoryDirNode()
+        self._base_url = base_url
+        setting_changed.connect(self._clear_cached_properties)
+
+    @cached_property
+    def base_url(self):
+        if self._base_url is not None and not self._base_url.endswith("/"):
+            self._base_url += "/"
+        return self._value_or_setting(self._base_url, settings.MEDIA_URL)
+
+    def _open(self, name, mode="rb"):
+        create_if_missing = "w" in mode
+        file_node = self._root.resolve(
+            name, create_if_missing=create_if_missing, leaf_cls=InMemoryFileNode
+        )
+        return file_node.open(mode)
+
+    def _save(self, name, content):
+        file_node = self._root.resolve(
+            name, create_if_missing=True, leaf_cls=InMemoryFileNode
+        )
+        fd = None
+
+        for chunk in content.chunks():
+            if fd is None:
+                mode = "wb" if isinstance(chunk, bytes) else "wt"
+                fd = file_node.open(mode)
+            fd.write(chunk)
+
+        file_node.modified_time = now()
+        return name
+
+    def delete(self, name):
+        path, filename = os.path.split(name)
+        dir_node = self._root.resolve(path)
+        if dir_node is None:
```

Comment:
Coverage says:
"line 211 didn't jump to line 212, because the condition on line 211 was never true"